src/test/java/org/springframework/samples/petclinic/cache/HybridCacheIntegrationTest.java

```diff
@@ -0,0 +1,440 @@
+        return Stream.of(
+            Arguments.of("vets", "/api/vets", "vets", "\"id\":1", "\"id\":2"),
+            Arguments.of("owners", "/api/owners", "owners", "\"id\":1", "\"id\":2"),
+            Arguments.of("pets", "/api/pets", "pets", "\"id\":1", "\"id\":2"),
+            Arguments.of("petTypes", "/api/pettypes", "petTypes", "\"id\":1", "\"id\":2"),
+            Arguments.of("specialties", "/api/specialties", "specialties", "\"id\":1", "\"id\":2"),
+            Arguments.of("visits", "/api/visits", "visits", "\"id\":1", "\"id\":2")
+        );
+    }
+
+    /**
+     * Test data provider for parameterized cache tests (single entity endpoints)
+     */
+    static Stream<Arguments> singleEntityCacheTestData() {
+        return Stream.of(
+            Arguments.of("vets", "/api/vets/1", "vetById", "\"id\":1", "Test"),
+            Arguments.of("owners", "/api/owners/1", "ownerById", "\"id\":1", "John"),
+            Arguments.of("pets", "/api/pets/1", "petById", "\"id\":1", "Fluffy"),
+            Arguments.of("petTypes", "/api/pettypes/1", "petTypeById", "\"id\":1", "Cat"),
+            Arguments.of("specialties", "/api/specialties/1", "specialtyById", "\"id\":1", "Dentistry"),
+            Arguments.of("visits", "/api/visits/1", "visitById", "\"id\":1", "Regular checkup")
+        );
+    }
+
+    @ParameterizedTest(name = "testHybridCacheCoordination_{0}")
+    @MethodSource("cacheTestData")
+    void testHybridCacheCoordination(String entityName, String endpoint, String cacheName, String expectedId1, String expectedId2) {
+        // 1. Seed through REST to populate both cache layers
+        ResponseEntity<String> response1 = restTemplate.getForEntity(endpoint, String.class);
+        assertThat(response1.getStatusCode()).isEqualTo(HttpStatus.OK);
+
+        // Verify L2 cache is populated in Redis
+        assertThat(findRedisKeysForCache(cacheName)).isNotEmpty();
+
+        // 3. Second call - should be faster due to caching
+        ResponseEntity<String> response2 = restTemplate.getForEntity(endpoint, String.class);
+        assertThat(response2.getStatusCode()).isEqualTo(HttpStatus.OK);
+        // Parse the JSON response to check it contains expected entities
+        assertThat(response2.getBody()).contains(expectedId1).contains(expectedId2);
+
+        // 4. Clear L1 cache only
+        clearL1Cache(cacheName);
+
+        // 5. Third call - should work with L2 but still succeed
+        clearAllRepositoryInvocations();
+        ResponseEntity<String> response3 = restTemplate.getForEntity(endpoint, String.class);
+        assertThat(response3.getStatusCode()).isEqualTo(HttpStatus.OK);
+        verifyNoRepositoryInteractions();
+
+        assertThat(findRedisKeysForCache(cacheName)).isNotEmpty();
+
+        // 6. Fourth call - should be cached again
+        ResponseEntity<String> response4 = restTemplate.getForEntity(endpoint, String.class);
+        assertThat(response4.getStatusCode()).isEqualTo(HttpStatus.OK);
```

Comment:
We do not really assert here anything meaningfull.
We probably should assert that here is no call to redis cache in this situation.
Not sure if it could be done easily.
May be it's enough to check that if we clear l2 here l1 there will be no interaction with database